dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+        {
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                nullptr,
+                channelRequestResult,
+                status);
+        }
+
+        co_return channelResult;
+    }
+
+    PushNotificationRegistrationToken PushNotificationManager::RegisterActivator(PushNotificationActivationInfo const& details)
+    {
+        winrt::guid taskClsid = details.TaskClsid();
+        DWORD cookie = 0;
+        BackgroundTaskRegistration registeredTask = nullptr;
+
+        if (WI_IsFlagSet(details.Kind(), PushNotificationRegistrationKind::PushTrigger))
+        {
+            bool taskRegistered = false;
+            for (auto task : BackgroundTaskRegistration::AllTasks())
+            {
+                if (task.Value().Name() == backgroundTaskName)
+                {
+                    taskRegistered = true;
+                    break;
+                }
+            }
+
+            if (!taskRegistered)
+            {
+                BackgroundTaskBuilder builder;
+                builder.Name(backgroundTaskName);
+
+                PushNotificationTrigger trigger{};
+                builder.SetTrigger(trigger);
+
+                if (taskClsid != winrt::guid())
+                {
+                    auto builder5 = builder.try_as<winrt::IBackgroundTaskBuilder5>();
+                    if (IsPackagedProcess() && builder5)
+                    {
+                        builder5.SetTaskEntryPointClsid(taskClsid);
+                        winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
+                        for (auto condition : conditions)
+                        {
+                            builder.AddCondition(condition);
+                        }
+                    }
+                }
+
+                registeredTask = builder.Register();
+            }
+        }
+
+        if (WI_IsFlagSet(details.Kind(), PushNotificationRegistrationKind::ComActivator))
+        {
+            if (taskClsid != winrt::guid())
+            {
+                {
+                    auto lock = g_lock.lock();
+                    // Define handle that will be set during background task execution
+                    g_waitHandleForArgs = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+                }
+
+                THROW_IF_FAILED(::CoRegisterClassObject(
+                    taskClsid,
+                    winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                    CLSCTX_LOCAL_SERVER,
+                    REGCLS_MULTIPLEUSE,
+                    &cookie));
+            }
+        }
+        return PushNotificationRegistrationToken{ cookie, registeredTask };
+    }
+
+    void PushNotificationManager::UnregisterActivator(PushNotificationRegistrationToken const& token, PushNotificationRegistrationKind const& kind)
```

Comment:
Why does the app have to pass back in the same `kind` that it passed the first time? What if they get confused and pass the wrong `kind` the second time? Shouldn't we remember the `kind` inside the `PushNotificationRegistrationToken`?

Reply:
This is a tricky one. We also want to allow the developer to delete the CLSID from BackgroundRegistration and not just unregister the COM activator in some cases. Deleting the CLSID from BackgroundRegistration is a corner case which implies that the app will never receive Push Notifications again.

If we were to simply use the token, we would not be able to account for this corner case scenario.